rust/kernel/spi.rs

```diff
@@ -0,0 +1,222 @@
+///     /// macro. For example, here our Driver will use the Kernel's shutdown method.
+///     fn probe(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+///
+///     fn remove(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! declare_spi_methods {
+    () => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::USE_NONE;
+    };
+    ($($method:ident),+) => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::ToUse {
+            $($method: true),+,
+            ..$crate::spi::USE_NONE
+        };
+    };
+}
+
+impl DriverRegistration {
+    fn new(this_module: &'static crate::ThisModule, name: &'static CStr) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            spi_driver: bindings::spi_driver::default(),
+        }
+    }
+
+    // FIXME: Add documentation
+    pub fn new_pinned<T: SpiMethods>(
+        this_module: &'static crate::ThisModule,
+        name: &'static CStr,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(this_module, name))?);
+
+        registration.as_mut().register::<T>()?;
+
+        Ok(registration)
+    }
+
+    unsafe extern "C" fn probe_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        match T::probe(SpiDevice::from_ptr(spi_dev)) {
+            Ok(_) => 0,
+            Err(e) => e.to_kernel_errno(),
+        }
+    }
+
+    unsafe extern "C" fn remove_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        match T::remove(SpiDevice::from_ptr(spi_dev)) {
+            Ok(_) => 0,
+            Err(e) => e.to_kernel_errno(),
+        }
+    }
+
+    unsafe extern "C" fn shutdown_wrapper<T: SpiMethods>(spi_dev: *mut bindings::spi_device) {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        T::shutdown(SpiDevice::from_ptr(spi_dev))
+    }
+
+    // FIXME: Add documentation
+    pub fn register<T: SpiMethods>(self: Pin<&mut Self>) -> Result {
+        fn maybe_get_wrapper<F>(vtable_value: bool, func: F) -> Option<F> {
+            match vtable_value {
+                false => None,
+                true => Some(func),
+            }
+        }
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver.driver.name = this.name.as_ptr() as *const c_types::c_char;
+        this.spi_driver.probe =
+            maybe_get_wrapper(T::TO_USE.probe, DriverRegistration::probe_wrapper::<T>);
```

Comment:
```rust
T::TO_USE.probe.then(|| DriverRegistration::probe_wrapper::<T>)
```

Reply:
Oh nice, we did not know about these methods. I'm patching it, and `then` does not write since this is not the correct type for the underlying wrapper. However, `then_some` is exactly the same behavior. But it's an unstable feature and needs to be added as a `#![feature]` in `kernel/lib.rs`. I'm not sure how you guys feel about that?

Reply:
`foo.then_some(bar)` is basically just `foo.then(|| bar)`. Did you wrap the expression inside a closure (`||`)?

Reply:
Yes, we tried using your suggestion exactly:

```rust
error[E0308]: mismatched types
   --> rust/kernel/spi.rs:149:33
    |
149 |           this.spi_driver.probe = T::TO_USE
    |  _________________________________^
150 | |             .probe
151 | |             .then(|| DriverRegistration::probe_wrapper::<T>);
    | |____________________________________________________________^ expected fn pointer, found fn item
    |
    = note: expected enum `Option<unsafe extern "C" fn(_) -> _>`
               found enum `Option<unsafe extern "C" fn(_) -> _ {DriverRegistration::probe_wrapper::<T>}>`
```

Reply:
Since `then_some` is simpler and simply return a generic option, it works fine. `then` is constrained to `FnOnce` functions

Reply:
Ah, I see, the issue is related to type inference inhibiting coercion. Unfortunately I think then_some will not be stabilized as it is so print it's better not to use it.

You can add an explicit cast to address the missing coercion, but maybe it's easier to just use if expression in this case.

Reply:
Out of curiosity, why do you think `then_some` will not be stabilized as it is? I've read through the tracking issues for it and it seems that the main detractor against the function agreed to keep it, but I might have missed some things :D

Reply:
From the tracking issue it seems that `then_some` isn't considered as a good name, so it'll likely change in the future.

TBH this doesn't look too bad
```rust
this.spi_driver.probe = if T::TO_USE.probe { Some(DriverRegistration::probe_wrapper::<T>) } else { None };
```

Reply:
Yeah, it's just a bit repetitive over the three fields, especially since it gets split up over 4 lines with rustfmt. This is what we had before basically reimplementing `then_some()` and calling it `maybe_get_wrapper()`

Reply:
Yeah, I do see that rustfmt dislikes having an if expression on a single line. What's your opinion on an explicit cast then?
```rust
this.spi_driver.probe = T::TO_USE.probe.then(|| DriverRegistration::probe_wrapper::<T> as _)
```

Reply:
I think this line is quite unclear... This is why I think `then_some` is great, it's super simple and easy to understand. I'm unsure that the API will ever change, and if it does there'll probably be a transition period where it's deprecated before being removed completely, so there will be time to fix it. If the API gets completely removed, then yeah I think using `then` is fine. However, I do not believe it will get removed so I'm more in favor of using `then_some` or the future name it will have. What do you think?

Reply:
There is no stability guarantee about nightly features at all; it's pretty clear to me that the Rust lib team is not happy with the name `then_some`, so that's could be changed in any nightly release; and it will cause immediate breakage for those who use latest nightly.

That's why nightly features should be avoided unless we have no alternatives, or that certain feature is on the track of stabilisation.

Reply:
Oh I didn't know about this, nor did I think hard enough! My bad. I'll apply your suggestion with the explicit cast! Thanks for taking the time

Reply:
Fixed in ed2cf7d! Thanks again